### ISSUE TITLE
Dependencies alias require the repo to be added

### ISSUE THREAD

**Issue by rocketspacer** [helm/helm#7710]
Output of `helm version`:            
```
version.BuildInfo{Version:"v3.1.1", GitCommit:"afe70585407b420d0097d07b21c47dc511525ac8", GitTreeState:"clean", GoVersion:"go1.13.8"}
```

Output of `kubectl version`:
```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.3", GitCommit:"06ad960bfd03b39c8310aaf92d1e7c12ce618213", GitTreeState:"clean", BuildDate:"2020-02-13T18:08:14Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.9-eks-502bfb", GitCommit:"502bfb383169b124d87848f89e17a04b9fc1f6f0", GitTreeState:"clean", BuildDate:"2020-02-07T01:31:02Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
```

Cloud Provider/Platform (AKS, GKE, Minikube etc.):  **AWS EKS**

I use fully qualified url for my Chart dependencies.

This works fine when I do `helm dep build`
```yaml
# Chart.yaml
dependencies:
  - name: argo-cd
    version: 1.8.0
    repository: https://argoproj.github.io/argo-helm
```
![image](https://user-images.githubusercontent.com/17026876/75619480-3421e200-5bb7-11ea-8b25-795fd2ebb18c.png)

However, whenever I use **alias**, like:
```yaml
# Chart.yaml
dependencies:
  - name: argo-cd
    version: 1.8.0
    repository: https://argoproj.github.io/argo-helm
    alias: argocd
```
`helm dep build` then require the repo to be added.
![image](https://user-images.githubusercontent.com/17026876/75619386-1d2ec000-5bb6-11ea-804c-c03b789b6f6f.png)

**Comment by mattfarina** [comment]
I just looked at the docs and this can be a bit confusing. We either need to clean up the docs or the feature.

You can add a repo using the `helm repo add` command with a name. The `alias` here enables you to refer to that name.

**Comment by mattfarina** [comment]
@rocketspacer from your understanding, how did you expect the feature to work? I'm trying to make sure I know so we can help close any gaps.

**Comment by rocketspacer** [comment]
@mattfarina The behaviour I'm expecting is:
If I specify full repo url for the dependency it shouldn't ask for the repo to be added. And this behaviour should be consistent whether I use the alias feature or not.

**Comment by philomory** [comment]
Yeah, the alias in `Chart.yaml` is supposed to be an alias for the resulting chart (i.e. the name of the subdirectory inside the `charts/` directory), it's not supposed to have anything to do with repository aliases that you get from `helm repo add`; it's worth noting that this already works correctly in `helm dependency update` even if you've ever called `helm repo add` at all.

**Comment by lindhe** [comment]
I was also surprised by this behavior. Since `helm dependency build --help` says "If no lock file is found, 'helm dependency build' will mirror the behavior of 'helm dependency update'" I expected it to work without adding the dependencies first, since `helm dependency update .` works perfectly.

**Comment by technosophos** [comment]
Just to be clear -- all of you are seeing this behavior _only_ when an `alias` is added?

**Comment by lindhe** [comment]
Good that you asked. No, I don't use alias and still get that behavior. Sorry, missed that part the first time I wrote.

**Comment by kristinnardal2** [comment]
I work with @lindhe and will give some details as he hasn't had the time to yet.
The problem we have is I think very much related to the this issue but doesn't have anything to do with aliases. I have created a github [repo](https://github.com/kristinnardal2/example-helm-repo) to be able to reproduce this.

Our problem is that `helm dep build` requires that the repository has to have been added with `helm repo add` for it to work. This is not the case when running `helm dep update`.

I think the solution to both of these problems would be to not require that the repository has been added, just like Helm already does with `helm dep update`.

**Comment by kristinnardal2** [comment]
The command is failing on a call to hasAllRepos

```bash
$ helm dep build --debug
Error: no repository definition for https://charts.bitnami.com/bitnami. Please add the missing repos via 'helm repo add'
helm.go:84: [debug] no repository definition for https://charts.bitnami.com/bitnami. Please add the missing repos via 'helm repo add'
helm.sh/helm/v3/pkg/downloader.(*Manager).hasAllRepos
	/home/circleci/helm.sh/helm/pkg/downloader/manager.go:414
helm.sh/helm/v3/pkg/downloader.(*Manager).Build
	/home/circleci/helm.sh/helm/pkg/downloader/manager.go:117
main.newDependencyBuildCmd.func1
	/home/circleci/helm.sh/helm/cmd/helm/dependency_build.go:68
github.com/spf13/cobra.(*Command).execute
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:842
github.com/spf13/cobra.(*Command).ExecuteC
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950
github.com/spf13/cobra.(*Command).Execute
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
main.main
	/home/circleci/helm.sh/helm/cmd/helm/helm.go:83
runtime.main
	/usr/local/go/src/runtime/proc.go:203
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1357
```

This function used to also be called from `helm dep update` but was removed from it in https://github.com/helm/helm/commit/6c6530c1b6a8111a1a1acce2628f24b3d4ff087b

Maybe we can just remove this function completely.

**Comment by kristinnardal2** [comment]
I've created a pull request to remove the hasAllRepos function. Local testing shows that this fixes the issue I have seen, I suspect that this also fixes the issue originally described in this issue.

**Comment by mattfarina** [comment]
@kristinnardal2 When you do a `helm dep build` we want to make sure it's pulling from a repository you have chosen to add. This is with security in mind. We don't want to fetch dependencies from unknown sources. That is the reason it wants them to be added.

**Comment by mattfarina** [comment]
@rocketspacer I am unable to reproduce. You're saying that when `alias` is set it asks you to add the repo but not when it is not set. Is that right?

**Comment by rocketspacer** [comment]
@mattfarina Yes I only experienced this with `alias`

**Comment by philomory** [comment]
This appears to be fixed in 3.4.1.

**Comment by lindhe** [comment]
I still think this needs some love.
